lib/EmitVerilog/EmitVerilog.cpp

```diff
@@ -739,32 +739,51 @@ class ExprEmitter : public ExprVisitor<ExprEmitter, SubExprInfo>,
   SubExprInfo emitBitSelect(Value operand, unsigned hiBit, unsigned loBit);
 
   SubExprInfo emitBinary(Operation *op, VerilogPrecedence prec,
-                         const char *syntax, bool hasStrictSign = false,
-                         bool opForceSign = false);
+                         const char *syntax, bool opForceSign = false);
 
   SubExprInfo emitVariadic(Operation *op, VerilogPrecedence prec,
-                           const char *syntax, bool hasStrictSign = false,
-                           bool opForceSign = false);
+                           const char *syntax, bool opForceSign = false);
 
-  SubExprInfo emitRTLSignedVariadic(Operation *op, VerilogPrecedence prec,
-                                    const char *syntax);
-
-  /// Emit the specified subexpression in a context where the sign matters,
-  /// e.g. for a less than comparison or divide.
-  SubExprInfo emitSignedBinary(Operation *op, VerilogPrecedence prec,
-                               const char *syntax) {
-    return emitBinary(op, prec, syntax, /*hasStrictSign:*/ true);
-  }
   /// Emit the specified subexpression in a context where the sign matters,
   /// e.g. for a less than comparison or divide.
   SubExprInfo emitRTLSignedBinary(Operation *op, VerilogPrecedence prec,
                                   const char *syntax) {
-    return emitBinary(op, prec, syntax, /*hasStrictSign:*/ true,
+    return emitBinary(op, prec, syntax,
                       /*opForceSign*/ true);
   }
   SubExprInfo emitUnary(Operation *op, const char *syntax, bool forceUnsigned) {
+    bool opSigned = getSignednessOf(op->getOperand(0).getType()) == IsSigned;
+    auto width = getBitWidthOrSentinel(op->getResult(0).getType());
+    auto opWidth = getBitWidthOrSentinel(op->getOperand(0).getType());
+
     os << syntax;
-    auto signedness = emitSubExpr(op->getOperand(0), Unary).signedness;
+    if (opSigned)
+      os << "$signed(";
```

Comment:
Won't this lead to emitting literals in a $signed() block?  We'd like to put the sign on the literal itself.  This seems like a bit that should get passed down into emitSubExpr.